dashboard/app/assets/javascripts/header.js.erb

```diff
@@ -2,110 +2,575 @@
+     *   currentLevelIndex: number,
+     *   levels: Array.<{
+     *     displayText: string,
+     *     status: string,
+     *     unplugged: boolean,
+     *     assessment: boolean
+     *   }>
+     * }}
+     */
+  HeaderStore = (function(_super) {
+    __extends(HeaderStore, _super);
+
+    // Must have a constructor
+    function HeaderStore() {
+      return HeaderStore.__super__.constructor.apply(this, arguments);
+    }
+
+    HeaderStore.prototype.resolveUrl = function(args, dataType) {
+      return "/popup/progress";
+    }
+
+    HeaderStore.prototype.resolveFilename = function(args, dataType) {
+      return "/progress_" + args.script_name + "_" + args.stage_id + "_" + args.level_id + "." + dataType;
+    }
+
+    return HeaderStore;
+  })(UIStore);
+
+
+
+  // ScriptStore: load the script progress (used in the header dropdown)
+  ScriptStore = (function(_super) {
+    __extends(ScriptStore, _super);
+
+    // Must have a constructor
+    function ScriptStore() {
+      return ScriptStore.__super__.constructor.apply(this, arguments);
+    }
+
+    ScriptStore.prototype.resolveUrl = function(args, dataType) {
+      return "/popup/stats";
+    }
+
+    ScriptStore.prototype.resolveFilename = function(args, dataType) {
+      return "/script_" + args.script_id + "." + dataType;
+    }
+
+    return ScriptStore;
+  })(UIStore);
+
+
+
+  // class HeaderProgress extends UIComponent
+  HeaderProgress = (function(_super) {
+    __extends(HeaderProgress, _super);
+
+    // Must have a constructor
+    function HeaderProgress($el, props) {
+      HeaderProgress.__super__.constructor.apply(this, arguments);
+
+      this.onModalClick = __bind(this.onModalClick, this);
+
+      return this;
+    }
+
+    HeaderProgress.prototype.componentDidMount = function() {
+      var _this = this;
+
+      // Create a new HeaderStore and bind it to my state.
+      // TODO: Is this the right place to put this?
+      // It guarantees there will be at least one render cycle before we try to load the header.
+      this.headerStore = new HeaderStore();
+      this.headerStore.notify( __bind(this.replaceState, this) );
+
+      this.scriptStore = new ScriptStore();
+      this.scriptStore.notify(function(html) {
+        _this.setState({
+          stageHtml: html
+        });
+      })
+
+      // TODO: Handle props changing.
+      if (this.props.script_name)
+        this.headerStore.load( this.props );
+
+      // Create tooltips for the free-play icons
+      // TODO: Might be better to initialize this component elsewhere
+      $('.level_free_play').qtip({
+        content: {
+          attr: 'title'
+        },
+        position: {
+          my: 'top center',
+          at: 'bottom center'
         }
       });
-    } else if (trophiesClicked) {
-      jumpToTrophies();
-    }
-  }
-  function jumpToTrophies() {
-    window.scrollTo(0, +$('#trophies').offset().top);
-  }
-};
+    }
+
+
+    HeaderProgress.prototype.render = function() {
```

Comment:
One significant difference between my thing and React is that React simply *returns* the new node structure, where mine adds it to the element.

This is significant because it means React render calls are a simple function that can be tested in isolation against whatever state you want, without having to instantiate the entire environment. Mine can't do that.